// react_cli/templates/webpack.config.hot.js
/* global module, __dirname */

var path = require('path');
var webpack = require('webpack');

module.exports = {
    devtool: 'cheap-module-eval-source-map',
    entry: {
        index: ['webpack-hot-middleware/client', './src/containers/IndexExt'],
        login: ['webpack-hot-middleware/client', './src/containers/UserLogin']
    },
    output: {
        path: path.join(__dirname, 'src', 'dist'),
        filename: '[name].js',
        publicPath: '/static/'
    },
    plugins: [
        new webpack.optimize.OccurenceOrderPlugin(),
        new webpack.HotModuleReplacementPlugin(),
        new webpack.NoErrorsPlugin(),
        new webpack.DefinePlugin({
            "process.env": {
                NODE_ENV: JSON.stringify("production")
            }
        })
    ],
    module: {
        loaders: [
            {test: /\.js$/, loaders: ['babel']},
            {test: /\.css?$/, loaders: ['style', 'raw'], include: __dirname},
            {test: /\.jsx$/, loader: 'babel-loader!jsx-loader?harmony'},
            {test: /\.(woff|woff2)$/, loader: "url-loader?limit=10000&mimetype=application/font-woff"},
            {test: /\.ttf$/, loader: "file-loader"},
            {test: /\.eot$/, loader: "file-loader"},
            {test: /\.svg$/, loader: "file-loader"}
        ]
    }
};
